akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.routing
+
+import akka.actor.typed.ActorRef
+import akka.annotation.InternalApi
+import akka.dispatch.forkjoin.ThreadLocalRandom
+
+/**
+ * Kept in the behavior, not shared between instances, meant to be stateful.
+ *
+ * INTERNAL API
+ */
+@InternalApi
+sealed private[akka] trait RoutingLogic[T] {
+  /**
+   * @param routees available routees, will contain at least one element. Must not be mutated by select logic.
+   */
+  def selectRoutee(): ActorRef[T]
+
+  /**
+   * Invoked an initial time before `selectRoutee` is ever called and then every time the set of available
+   * routees changes.
+   *
+   * @param newRoutees The updated set of routees. For a group router this could be empty, in that case
+   *                   `selectRoutee()` will not be called before `routeesUpdated` is invoked again with at
+   *                   least one routee. For a pool the pool stops instead of ever calling `routeesUpdated`
+   *                   with an empty list of routees.
+   */
+  def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object RoutingLogics {
+
+  final class RoundRobinLogic[T] extends RoutingLogic[T] {
+
+    private var currentRoutees: Array[ActorRef[T]] = _
+
+    private var nextIdx = 0
+
+    def selectRoutee(): ActorRef[T] = {
+      if (nextIdx >= currentRoutees.length) nextIdx = 0
+      val selected = currentRoutees(nextIdx)
+      nextIdx += 1
+      selected
+    }
+
+    override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
+      // make sure we keep a somewhat similar order so we can potentially continue roundrobining
+      // from where we were unless the set of routees completely changed
+      // Also, avoid putting all entries from the same node next to each other in case of cluster
+      val sortedNewRoutees = newRoutees.toArray.sortBy(ref ⇒ (ref.path.toStringWithoutAddress, ref.path.address))
+
+      if (currentRoutees ne null) {
+        val firstDiffIndex = {
+          var idx = 0
+          while (idx < currentRoutees.length &&
+            idx < sortedNewRoutees.length &&
+            currentRoutees(idx) == sortedNewRoutees(idx)) {
+            idx += 1
+          }
+          idx
+        }
+        if (nextIdx > firstDiffIndex) nextIdx -= 1
+      }
+      currentRoutees = sortedNewRoutees
+    }
+  }
+
+  final class RandomLogic[T] extends RoutingLogic[T] {
+
+    private var currentRoutees: Array[ActorRef[T]] = _
```

Comment:
`var` in a regular class with no guard :( Is this needing perhaps a `self: ActorRef[T] =>` or something?

Reply:
Or maybe this is just set and accessed actually.

Reply:
It's not mixed into the behavior, the actual pool and group actors guarantee that it is safe to keep state. That it is provided through a factory guarantees no shared instances.

Reply:
I should add, the `RoutingLogic` is not possible to implement without keeping state, so that is very much on purpose. Note that it is not public API for now (and not really planning to make it public either, we'll see)